libcraft/items/src/item_stack.rs

```diff
@@ -94,62 +92,72 @@ impl ItemStack {
     /// Returns whether the given `ItemStack` has
     /// the same count as (but not necessarily the same
     /// type as) `self`.
+    #[must_use]
     pub fn has_same_count(&self, other: &Self) -> bool {
         self.count == other.count
     }
 
     /// Returns whether the given `ItemStack` has the same
     /// type and count as (but not necessarily the same meta
     /// as) `self`.
+    #[must_use]
     pub fn has_same_type_and_count(&self, other: &Self) -> bool {
         self.item == other.item && self.count == other.count
     }
 
     /// Returns whether the given `ItemStack` has
     /// the same type and damage as `self`.
+    #[must_use]
     pub fn has_same_type_and_damage(&self, other: &Self) -> bool {
         self.item == other.item && self.has_same_damage(other)
     }
 
     /// Returns the item type for this `ItemStack`.
-    pub fn item(&self) -> Item {
+    #[must_use]
+    pub const fn item(&self) -> Item {
         self.item
     }
 
     /// Returns the number of items in this `ItemStack`.
-    pub fn count(&self) -> u32 {
+    #[must_use]
+    pub const fn count(&self) -> u32 {
         self.count.get()
     }
 
     /// Adds more items to this `ItemStack`. Returns the new count.
+    /// # Errors
+    /// Returns `ExceedsStackSize` when the combined amount of items is greater than the stack size.
+    /// If the new item count cannot be represented as `i32`, returns `ClientOverflow`.
     pub fn add(&mut self, count: u32) -> Result<u32, ItemStackError> {
         self.set_count(self.count.get() + count)
     }
 
     /// Adds more items to this `ItemStack`. Does not check if the
     /// addition will make the count to be greater than the
     /// stack size. Does not check count overflows. Returns the new count.
+    /// # Panics
+    /// Panics if the new item count is greater than the stack size.
     pub fn unchecked_add(&mut self, count: u32) -> u32 {
         self.count = NonZeroU32::new(self.count.get() + count).unwrap();
         self.count.get()
     }
 
-    /// Removes some items from this `ItemStack`.
+    /// Removes some items from this `ItemStack` and return the new item count.
+    /// # Errors
+    /// Returns `NotEnoughItems` if the new stack would be empty.
+    #[allow(clippy::missing_panics_doc)]
     pub fn remove(&mut self, count: u32) -> Result<u32, ItemStackError> {
         if self.count.get() <= count {
```

Comment:
Maybe use checked subtraction.

Reply:
`NonZeroU32` doesn't implement `checked_sub`, can you explain what you meant?

Reply:
Instead of check that `count.get() >= count` and then doing the subtraction. Ie
```rust
self.count = self
  .count
  .get()
  .checked_sub(count)
  .and_then(NonZeroU32::new)
  .ok_or(ItemStackError::NotEnoughItems)?;
Ok(self.count.get())
```
Roughly

Reply:
But shouldn't an itemstack be removed when we remove exactly the amount of items it has? With this code we would then fail instead of delete the itemstack

Reply:
The thing is, an itemstack cannot have 0 items and I don't think having two separate errors is neccessary.